### PR TITLE
Filter new arrived channel before merge with current state on SET_CHA…

### PR DESCRIPTION
**Pull request by AleshaOleg**
…NNELS action

### CODE REVIEW

**Comment by AleshaOleg** [comment]
Previously this action (`SET_CHANNELS`) works ok with `return [...action.payload]`. But then I found a bug which prevents loading user with whom current user didn't have conversation yes. I decided to return `state + payload`, it's worked ok because an app has two async function which loads separately at initial loading:
1) https://github.com/ubergrape/grape-web-client/blob/master/src/app/subscribe.js#L55, when connection to web-socket finished, and app received client id.
2) https://github.com/ubergrape/grape-web-client/blob/master/src/containers/router/Router.js#L51
And when route changed (obviously it's happening while new page loading).
So, we don't know which action will be loaded first.

But, there's a case, when connection to web-socket closed (didn't receive pong, or error, etc..). The app will try to reconnect to the socket, and if it receives a new client id, initial loading will be triggered again (see the first point of initial loading process).

In this case, we will have a two copy of all our channels (because we just add the payload to state on `SET_CHANNELS`). That's why I decided to remove channels that already exist in the state, and replace them with fresh from payload? What do you think @nikgraf?

File: src/reducers/channels.js [hunk 1]
```diff
@@ -59,7 +62,7 @@ export default function reduce(state = initialState, action) {
         // we have to ensure that user isn't joined already.
         // https://github.com/ubergrape/chatgrape/issues/3804
         users: includes(users, userId) ? users : [...users, userId],
-        joined: isCurrentUser ? true : channel.joined
+        joined: isCurrentUser ? true : channel.joined,
```
Comment (glennreyes) [inline src/reducers/channels.js]: Can be simplified to `isCurrentUser || channel.joined`

**Comment by AleshaOleg** [comment]
@nikgraf description and tests added. Never wrote tests for reducers before, would be good if you will review it very attentively!

File: src/reducers/channels.js (hunk 1 above)
Comment (AleshaOleg) [reply src/reducers/channels.js]: @glennreyes nice catch! thanks:)

File: src/reducers/__tests__/index.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    ).toEqual([c1, c2])
+  })
+
+  it('should handle correct merge channels on SET_CHANNELS', () => {
+    expect(
+      channels([c1, c2], {
+        type: types.SET_CHANNELS,
+        payload: [c1, c3],
+      }),
+    ).toEqual([c2, c1, c3])
```
Comment (nikgraf) [inline src/reducers/__tests__/index.js]: Great test 👍

**Pull request merged**